hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/WritableKVStateBase.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.spi.state;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.*;
+
+/**
+ * A base class for implementations of {@link WritableKVState}.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+public abstract class WritableKVStateBase<K extends Comparable<K>, V>
+        extends ReadableKVStateBase<K, V> implements WritableKVState<K, V> {
+    /** A map of all modified values buffered in this mutable state */
+    private final Map<K, V> modifications = new HashMap<>();
+
+    /**
+     * Create a new StateBase.
+     *
+     * @param stateKey The state key. Cannot be null.
+     */
+    protected WritableKVStateBase(@NonNull final String stateKey) {
+        super(stateKey);
+    }
+
+    /**
+     * Flushes all changes into the underlying data store. This method should <strong>ONLY</strong>
+     * be called by the code that created the {@link WritableKVStateBase} instance or owns it. Don't
+     * cast and commit unless you own the instance!
+     */
+    public final void commit() {
+        for (final var entry : modifications.entrySet()) {
+            final var key = entry.getKey();
+            final var value = entry.getValue();
+            if (value == null) {
+                removeFromDataSource(key);
+            } else {
+                putIntoDataSource(key, value);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Clears the set of modified keys and removed keys. Equivalent semantically to a "rollback"
+     * operation.
+     */
+    @Override
+    public final void reset() {
+        super.reset();
+        modifications.clear();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Nullable
+    public final V get(@NonNull K key) {
+        // If there is a modification, then we've already done a "put" or "remove"
+        // and should return based on the modification
+        if (modifications.containsKey(key)) {
+            final var value = modifications.get(key);
+            super.markRead(key, value);
```

Comment:
Do we actually need to track the read keys? My understanding is, that this is needed only during prehandle, where we do not work with a `WritableKVState`.

Reply:
I am not sure we need to keep track of read keys in a WritableKVState, but I did so to faithfully implement the contract. I could remove this logic and throw an exception if you try to get the readKeys?

Reply:
Another option would be to remove the parent class. My impression is, that `ReadableKVStateBase` does not really provide a lot of functionality to this class, but adds complexity, because you have to think about `readKeys` all the time.

Reply:
I think I prefer to leave this as is, because we will probably need it. I know, YAGNI, but since it is already implemented I'd rather just let it go in. It enables us to actually support full writable states (that we don't commit) in pre-handle if we want to do it...